src/ghci/warning_tracker.rs

```diff
@@ -0,0 +1,335 @@
+//! Warning tracking and management for GHC diagnostics.
+//!
+//! This module provides functionality to track warnings across recompilations,
+//! managing the lifecycle of warnings as files are modified, added, or removed.
+
+use std::borrow::Borrow;
+use std::collections::{BTreeMap, BTreeSet};
+
+use crate::ghci::parse::GhcDiagnostic;
+use crate::ghci::CompilationLog;
+use crate::normal_path::NormalPath;
+
+/// Tracks warnings across recompilations, managing their lifecycle.
+///
+/// This tracker maintains an in-memory store of warnings per file and provides
+/// smart logic for updating warnings based on compilation results and file changes.
+pub struct WarningTracker {
+    /// Per-file warnings from GHC compilation, persisted across reloads.
+    warnings: BTreeMap<NormalPath, Vec<GhcDiagnostic>>,
+    /// Files that were directly changed in the current reload operation.
+    /// Used to distinguish between direct changes and dependency-driven recompilations.
+    current_changed_files: BTreeSet<NormalPath>,
+}
+
+impl WarningTracker {
+    /// Create a new warning tracker.
+    pub fn new() -> Self {
+        Self {
+            warnings: BTreeMap::new(),
+            current_changed_files: BTreeSet::new(),
+        }
+    }
+
+    /// Reset the list of files that were directly changed.
+    /// This should be called at the start of each reload operation.
+    pub fn reset_changed_files(&mut self) {
+        self.current_changed_files.clear();
+    }
+
+    /// Mark a file as having been directly changed.
+    /// This affects how warnings are managed for this file.
+    pub fn mark_file_changed(&mut self, path: NormalPath) {
+        self.current_changed_files.insert(path);
+    }
+
+    /// Update warnings from a compilation log.
+    ///
+    /// This method implements smart warning persistence logic:
+    /// - Files that were directly changed: always update warnings (or clear if none)
+    /// - Files that were recompiled due to dependencies: only update if warnings exist
+    /// - Files that weren't recompiled: keep existing warnings
+    pub fn update_warnings_from_log(&mut self, log: &CompilationLog) {
```

Comment:
I think this is the only method that uses `current_changed_files`, would it make sense to make that variable a parameter to this function instead of a field on the struct?

Reply:
I could go either way? Do you feel strongly one way or the other? 

My thinking here is that the struct keeps track of the things relevant for the feature: warnings and the files currently being changed (to determine which warnings are displayed). The latter _could_ be thought of as an implementation detail but I wasn't sure if we'd want to surface the file list elsewhere externally too in the future.